server/src-lib/Hasura/GraphQL/Parser/Schema.hs

```diff
@@ -0,0 +1,684 @@
+are happy to take advantage of this coincidence. -}
+
+-- | Evidence for '<:'.
+data k1 :<: k2 where
+  KRefl :: k :<: k
+  KBoth :: k :<: 'Both
+
+-- | 'Kind' subsumption. The GraphQL kind hierarchy is extremely simple:
+--
+-- >     Both
+-- >     /  \
+-- > Input  Output
+--
+-- Various functions in this module use '<:' to allow 'Both' to be used in
+-- places where 'Input' or 'Output' would otherwise be expected.
+class k1 <: k2 where
+  subKind :: k1 :<: k2
+instance k1 ~ k2 => k1 <: k2 where
+  subKind = KRefl
+instance {-# OVERLAPPING #-} k <: 'Both where
+  subKind = KBoth
+
+data Type k
+  = NonNullable (NonNullableType k)
+  | Nullable (NonNullableType k)
+
+instance Eq (Type k) where
+  (==) = eqType
+
+-- | Like '==', but can compare 'Type's of different kinds.
+eqType :: Type k1 -> Type k2 -> Bool
+eqType (NonNullable a) (NonNullable b) = eqNonNullableType a b
+eqType (Nullable    a) (Nullable    b) = eqNonNullableType a b
+eqType _               _               = False
+
+instance HasName (Type k) where
+  getName = getName . discardNullability
+
+instance HasDefinition (Type k) (TypeInfo k) where
+  definitionLens f (NonNullable t) = NonNullable <$> definitionLens f t
+  definitionLens f (Nullable t)    = Nullable <$> definitionLens f t
+
+discardNullability :: Type k -> NonNullableType k
+discardNullability (NonNullable t) = t
+discardNullability (Nullable t)    = t
+
+nullableType :: Type k -> Type k
+nullableType = Nullable . discardNullability
+
+nonNullableType :: Type k -> Type k
+nonNullableType = NonNullable . discardNullability
+
+data NonNullableType k
+  = TNamed (Definition (TypeInfo k))
+  | TList (Type k)
+
+instance Eq (NonNullableType k) where
+  (==) = eqNonNullableType
+
+toGraphQLType :: Type k -> GType
+toGraphQLType = \case
+  NonNullable t -> translateWith False t
+  Nullable    t -> translateWith True  t
+  where
+    translateWith nullability = \case
+      TNamed typeInfo -> TypeNamed (Nullability nullability) $ getName typeInfo
+      TList  typeInfo -> TypeList  (Nullability nullability) $ toGraphQLType typeInfo
+
+
+-- | Like '==', but can compare 'NonNullableType's of different kinds.
+eqNonNullableType :: NonNullableType k1 -> NonNullableType k2 -> Bool
+eqNonNullableType (TNamed a) (TNamed b) = liftEq eqTypeInfo a b
+eqNonNullableType (TList  a) (TList  b) = eqType a b
+eqNonNullableType _          _          = False
+
+instance HasName (NonNullableType k) where
+  getName (TNamed definition) = getName definition
+  getName (TList t)           = getName t
+
+instance HasDefinition (NonNullableType k) (TypeInfo k) where
+  definitionLens f (TNamed definition) = TNamed <$> f definition
+  definitionLens f (TList t)           = TList <$> definitionLens f t
+
+data InputObjectInfo = InputObjectInfo ~[Definition InputFieldInfo]
+-- Note that we can't check for equality of the fields since there may be
+-- circularity. So we rather check for equality of names.
```

Comment:
Doing this name-only comparison is a bit sketchy, and it’s part of the reason the memoization exists: the `dUnique` field is supposed to break cycles. That way, we can detect an inconsistent schema where the names are the same, but some other part of the structure is different.

However, the current algorithm isn’t clever enough to do this equality check. The issue is that we need full graph equality. We might have two values that each form a cycle with themselves, but are otherwise completely distinct. To illustrate what I mean, consider this pseudo-code:

```haskell
let xs = 1 : 2 : 3 : xs
    ys = 1 : 2 : 3 : ys
```

In this case, `xs == ys` should be `True`, even though they are two totally independent cycles.

Our current algorithm only takes advantage of `dUnique` by returning `True` if two `Definitions` compared for equality have the same `Unique`. That isn’t enough in an example like the above, since at no point will there be a truly shared `Unique`, the structures formed by the self-references are simply the same.

Now, I’m not actually sure if this scenario can ever happen in our code. Since we only install new `Unique`s on memoization, it seems like two cycles could only ever be equal if they truly share structure. But I’m not sure. It seems worth investigating, since doing it this way means we’re essentially not getting any value out of the `Unique`s and might as well not use them if we’re going to rely on `Name` equality here.

Reply:
I completely agree with the sketchiness. The code works this way to avoid actual infinite loops that we did encounter.

> Now, I’m not actually sure if this scenario can ever happen in our code. Since we only install new `Unique`s on memoization, it seems like two cycles could only ever be equal if they truly share structure.

If we only install `Unique`s on memoization, then I see one way in which testing equality on `Unique` will currently not suffice. We may have circular types defined in a remote schema. When those types are also used in a remote relationship, then, currently, the types in the remote schema are built in a separate `runSchemaT` from when those types in the remote relationship are built up. After all, the second pass is when we generate schema for database tables. And, currently, we do not re-use the type information from remote schemas when inserting remote relationships into database-type selection sets. But perhaps there is a way to do that.

Perhaps we can resolve this by leaving behind a more visible note on the justifications and sketchiness?

Reply:
In theory, implementing full graph equality is not hard: you just have to record the `Unique`s you’ve compared for equality so far, and if you see the same pair of `Unique`s again during the same traversal, you assume the values are equal (since that is the premise, and you’re searching for a counterexample). It seems worth giving it a try and seeing if that eliminates the need for the sketchiness?

Reply:
(For the record: I'm looking into implementing this. Alexis and I are now thinking that a fully correct implementation could be slightly more involved than this, but probably still doable.)